stress_schedule: Fix for !GIL + emit=native.

jepler opened the pull request:
### Summary

Ensure that zero-duration sleeps run scheduled callbacks.

Reduce a wait making the test run 8x faster locally, which hopefully also makes it reliable on ci.

Restore the test during ci as well as some asyncio tests.

Closes: #18415 

ping @projectgus 

### Testing

I ran the script manually locally.

### Trade-offs and Alternatives

We could delete stress_schedule, it's been unused in CI since 2020.

jepler committed:
stress_schedule: Decrease backoff time.

Locally, this changes the duration of the test from about 2.7s
to 0.3s.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
run-tests: Re-add stress_schedule.py.

This test was ignored since 2020, which hid a new bug in the
test for the native emitter added in 2024. Hopefully other
changes to the test will make it more reliable.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler commented:
I can look at it from that angle instead, thanks.

jepler committed:
unix: Handle pending events during sleep(0).

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

dpgeorge commented on extmod/modtime.c:
```diff
@@ -134,6 +134,8 @@ static mp_obj_t time_sleep_ms(mp_obj_t arg) {
     if (ms >= 0) {
         mp_hal_delay_ms(ms);
+    } else {
+        mp_handle_pending(true);
```
Hmm, I'm not sure this bit needs to be added.  There doesn't seem to be anything broken that needs this change.

dpgeorge replied:
I don't understand.

The existing logic here is `if (ms >= 0)`, so this change won't change anything for `time.sleep_ms(0)`.

jepler committed:
unix: Ensure mp_hal_delay_ms handles pending tasks.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
tests: Add schedule_sleep.py.

This variant of schedule.py explicitly calls a zero sleep.
The existing variant is kept to ensure the scheduler is called
between bytecodes.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
docs: Document that sleep(0), sleep_ms(0) run the scheduler.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
qemu: Fix delay_ms to call scheduler.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
modselect: Handle pending before entering poll.

This allows tests like asyncio_event_queue.py to succeed under
the native emitter when poll is enabled.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler commented:
I changed unix sleep_ms as requested and added to the documentation the guarantee (possibly not yet met in all ports!) that sleep(0) and sleep_ms(0) run scheduled tasks. I also found that some asyncio tests would run under the native emitter provided that `poll_poll_internal` C function called pending tasks.

Local testing showed that qemu's sleep_ms needed the added code to run callbacks for sleep_ms(0). Thus, it's likely that ALL ports should have their implementations checked & specifically the micropython/schedule.py test run under the native emitter.

 I'll check back and see how CI is doing.

dpgeorge commented on ports/unix/unix_mphal.c:
```diff
@@ -243,6 +243,7 @@ uint64_t mp_hal_time_ns(void) {
 #ifndef mp_hal_delay_ms
 void mp_hal_delay_ms(mp_uint_t ms) {
     mp_uint_t start = mp_hal_ticks_ms();
+    mp_handle_pending(true);
```
I think it would be better to duplicate it, because the existing one is testing that the VM calls the scheduler without explicit calls to sleep/poll/etc.

jepler committed:
tests: These tests no longer fail under emitter.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
stress_schedule: Remove inaccurate comment.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

dpgeorge commented on extmod/modselect.c:
```diff
@@ -537,6 +537,8 @@ MP_DEFINE_CONST_FUN_OBJ_3(poll_modify_obj, poll_modify);
 static mp_uint_t poll_poll_internal(uint n_args, const mp_obj_t *args) {
     mp_obj_poll_t *self = MP_OBJ_TO_PTR(args[0]);
 
+    mp_handle_pending(true);
```
I suggest moving this to the start of `poll_set_poll_until_ready_or_timeout()`, so that it's also called for `select.select()`.

jepler replied:
done.

dpgeorge commented on ports/unix/modtime.c:
```diff
@@ -95,6 +95,7 @@ static mp_obj_t mp_time_sleep(mp_obj_t arg) {
     tv.tv_sec = (suseconds_t)ipart;
     int res;
     while (1) {
+        mp_handle_pending(true);
```
This looks fine to me.

dpgeorge commented:
Thanks for updating.

The CI failures look unrelated to this PR's changes.  I'm guessing the Python version changed in the Windows CI and that led to settrace results differing?  Well, that will need to be investigated separately.

dpgeorge reviewed (approved):
This looks good now, thanks.

I also tested it on RP_PICO_W and ESP32_GENERIC_C3, running the tests in native mode, and they pass.
